tests/fe/fe_q_3rd_derivative_divergence_theorem.cc

```diff
@@ -0,0 +1,176 @@
+        {
+          deallog << i << ": ( ";
+          for (unsigned int d=0; d<dim; ++d)
+            deallog << cell->vertex(i)[d] << " ";
+          deallog << ")" << std::endl;
+        }
+
+      bool cell_ok = true;
+
+      for (unsigned int c=0; c<fe.n_components(); ++c)
+        {
+          FEValuesExtractors::Scalar single_component (c);
+
+          for (unsigned int i=0; i<fe_values.dofs_per_cell; ++i)
+            {
+              ss << "component=" << c
+                 << ", dof=" << i
+                 << std::endl;
+
+              Tensor<3,dim> bulk_integral;
+              for (unsigned int q=0; q<fe_values.n_quadrature_points; ++q)
+                {
+                  bulk_integral += fe_values[single_component].third_derivative(i,q) * fe_values.JxW(q);
+
+                  Tensor<3,dim> third_derivative = fe_values[single_component].third_derivative(i,q);
+                }
+
+              Tensor<3,dim> boundary_integral;
+              for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
+                {
+                  fe_face_values.reinit(cell,face);
+                  for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+                    {
+                      Tensor<2,dim> hessian = fe_face_values[single_component].hessian (i,q);
+                      Tensor<3,dim> hessian_normal_outer_prod;
+
+                      outer_product(hessian_normal_outer_prod, hessian, fe_face_values.normal_vector(q));
+                      boundary_integral += hessian_normal_outer_prod * fe_face_values.JxW(q);
+                    }
+                }
+
+              if ((bulk_integral-boundary_integral).norm_square() > tolerance * (bulk_integral.norm() + boundary_integral.norm()))
+                {
+                  deallog << "Failed:" << std::endl;
+                  deallog << ss.str() << std::endl;
+                  deallog << "    bulk integral=" << bulk_integral << std::endl;
+                  deallog << "boundary integral=" << boundary_integral << std::endl;
+                  deallog << "Error! difference between bulk and surface integrals is "
+                          << (bulk_integral-boundary_integral).norm_square()
+                          << " and greater than "
+                          << tolerance * (bulk_integral.norm() + boundary_integral.norm())
+                          << "!\n\n" << std::endl;
+                  cell_ok = false;
+                }
+
+              ss.str("");
+            }
+        }
+
+      deallog << (cell_ok? "OK: cell bulk and boundary integrals match...\n" : "Failed divergence test...\n") << std::endl;
+    }
+}
+
+
+
+template<int dim>
+void test_hyper_ball(const double tolerance)
+{
+  Triangulation<dim> tr;
+  GridGenerator::hyper_ball(tr);
+
+  static const HyperBallBoundary<dim> boundary;
+  tr.set_boundary (0, boundary);
+
+  tr.refine_global(1);
+
+  FE_Q<dim>  fe(3);
```

Comment:
I think it would be easiest if these other tests were kept in separate files, so that if one of the elements is patched to do something stupid, it is easy to find out which one it was.